toffy/qc_metrics_plots.py

```diff
@@ -0,0 +1,72 @@
+from ark.utils.load_utils import load_imgs_from_tree, load_imgs_from_dir
+import matplotlib.pyplot as plt
+import seaborn as sns
+import numpy as np
+from toffy import qc_comp
+from ark.utils import io_utils
+import pandas as pd
+
+def load_normal_fovs(data_dir, 
+	                 normal_tissues,
+                     channels=None,
+                     exclude_channels=None,
+                     img_sub_folder=None):
+    
+    normal_samples = io_utils.list_folders(dir_name=data_dir, substrs=normal_tissues)
+    data = load_imgs_from_tree(data_dir=data_dir, 
+	                    img_sub_folder=img_sub_folder, 
+	                    fovs=normal_samples, 
+	                    channels=channels)    
+    return data
+
+def qc_metrics_plots(data_dir, 
+	                 normal_tissues,
+                     channels=None,
+                     exclude_channels=None,
+	                 qc_metric="99.9th_percentile",
+	                 fig_dir=None):
+    
+    normal_samples = io_utils.list_folders(dir_name=data_dir, substrs=normal_tissues)
+
+    # need to make a nice pandas DF with columns: sample, channel, 99.9th value
+    for i in range(len(normal_tissues)):
+        data = load_normal_fovs(data_dir,
+                                normal_tissues=normal_tissues,
+                                channels=channels)
+        channels = list(data.channels.values)
+        if exclude_channels:
+            channels = [x for x in channels if x not in exclude_channels]
+        
+        
+        # For each tissue type, get all the samples
+        samples = [x for x in normal_samples if normal_tissues[i] in x]
+        plotting_df = pd.DataFrame(columns=["sample",'channel',"tma","99.9th_percentile"])
+        
+        for j in range(len(channels)):
+            # Channel is a gene
+            qc_metrics_per_channel = []
+
+            for k in range(len(samples)):
+                # go through all samples per gene. list should be of length==samples
+                tma = [x for x in samples[k].split("_") if "TMA" in x][0]
+                qc_metrics_per_channel += [[normal_tissues[i],
+                                           channels[j],
+                                           tma,
+                                           qc_comp.compute_99_9_intensity(data.loc[samples[k], 0:1023, 0:1023, channels[j]])]]
+            
+
+            plotting_df = plotting_df.append( pd.DataFrame(qc_metrics_per_channel,columns=plotting_df.columns) )
```

Comment:
I could put it into a separate function to make it more neat looking and parameterize the fig size ( less wide figure when plotting fewer markers etc. ). But  it is such a specific plot requiring a specific data structure and its only ~4 lines of plotting code i dont think it could be repurposed.

Reply:
I agree that the chances of this specific figure plotting code being reused are not that high. However, the benefit of making it more modular is that it makes it easier to read the code, as well as re-use it. 

If there is one function to creates the data frame, and a second function to plot the dataframe, then when someone is inspecting this code to figure out what it does, they'll see those two helper functions being called, giving a high-level understanding of the logic. Then, if they want to know how one part works, they can then look just at the relevant function. 

The other benefit is for changing this code in the future. If the functionality to extract the data is separate from the functionality to plot the data, if someone wants to add in a different plot, or save the data instead of plotting, it's very simple to make a new plotting function or saving function that gets called instead of this one.